### PR TITLE
use settings instead of config

### PR DESCRIPTION
**Pull request by JarbasAl**

### CODE REVIEW

**Comment by ChristopherRogers1991** [comment]
Hey,

Thanks for sending this in. I'm away from home for a couple days, but I'll take a look at this when I'm back home next week.

-Chris

**Comment by JarbasAI** [comment]
i also added a mark1 animation for the faceplate while waiting for results :)

its nice to have a visual cue that it is working and didnt just crash waiting for results

File: settingsmeta.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        "sections": [
+            {
+                "name": "Options",
+                "fields": [
+                    {
+                        "name": "host",
+                        "type": "text",
+                        "label": "host",
+                        "placeholder": "speedtest-po.vodafone.pt:8080",
+                        "value": "speedtest-po.vodafone.pt:8080"
```
Comment (ChristopherRogers1991) [inline settingsmeta.json:14]: We shouldn't be setting a default value for host; let's set the value to `""` and either remove the placeholder, or set it to something more generic like `hostname:port`.

With no host supplied, the SpeedTest class will find the "best" host (see https://github.com/fopina/pyspeedtest/blob/master/pyspeedtest.py#L67). This is the most familiar and robust behavior (when you go to speedtest.net, it selects the best host, and if the host goes down, our skill won't fail).

File: __init__.py [hunk 2]
```diff
@@ -150,16 +234,17 @@ def handle_speedtest_intent(self, message):
...
                    + "the download speed was " + download +
                    ", and the upload speed was " + upload)
 
     def stop(self):
-        """
-        Not implemented
-
-        """
-        pass
+        if self.playing:
```
Comment (ChristopherRogers1991) [inline __init__.py:243]: This doesn't seem to work as expected. Asking mycroft to stop in the middle of the speedtest doesn't seem to have any effect.

File: __init__.py [hunk 3]
```diff
@@ -140,6 +204,26 @@ def handle_speedtest_intent(self, message):
 
         """
         self.speak_dialog('start')
+        wait_while_speaking()
+
+        if not self.thread:
```
Comment (ChristopherRogers1991) [inline __init__.py:209]: I think this `handle_speedtest_intent` method is growing a bit too large, and taking on a bit too much. We should extract this `if` block into a function (something like `begin_animation` or `begin_eye_animation`). If you're up for it, lines 228 - 235 (the actual speedtest) could also be extracted into a `run_speedtest` function that returns a tuple with the results. (I like to aggressively refactor, to keep things neat).

File: __init__.py [hunk 4]
```diff
@@ -102,24 +106,84 @@ def attempt_three_times(function):
...
             msg = "Caught {error}: {message}. Retrying...".format(
                 error=type(e), message=e.message)
-            logger.warning(msg)
+            LOG.warning(msg)
+
+
+#######################################################################
+# Helpers for building the animation frames
+
+def animate(t, often, func, *args):
```
Comment (ChristopherRogers1991) [inline __init__.py:115]: I don't think this animation code should live here. This seems like it should be extracted into a utils module/class (and probably as a part of Mycroft core - seems silly to have each skill make it's own animation logic).

File: __init__.py [hunk 5]
```diff
@@ -102,24 +106,84 @@ def attempt_three_times(function):
...
+        self.playing = False
+        self.thread = None
+
+    def run(self):
+        """
+        animation thread while performing speedtest
+
+        """
+
+        while self.playing:
```
Comment (ChristopherRogers1991) [inline __init__.py:158]: Same note here about the animation - I don't think it belongs in this class.

File: __init__.py [hunk 6]
```diff
@@ -102,24 +106,84 @@ def attempt_three_times(function):
...
+def _get_time(often, t):
+    return often - t % often
 
 
 class SpeedTestSkill(MycroftSkill):
 
     def __init__(self):
         super(SpeedTestSkill, self).__init__(name="SpeedTestSkill")
         self.speedtest = None
+        if "host" not in self.settings:
```
Comment (ChristopherRogers1991) [inline __init__.py:142]: I think it would be better to just use `.get(<field>, <default_value>)` in `initialize`, rather than setting these values here.

**Review by ChristopherRogers1991 — CHANGES_REQUESTED** [review]
Hey @JarbasAl, thanks for submitting this. I think the settings code is basically ready to go (once we remove the default host), but I'd like to iterate more on the animations. If you'd like to split this into two PRs, with the settings and animation separated out, we can probably get the settings portion merged quickly, while we iterate on the animation.

@Sujan4k0 will also be taking a look, so we should have a third opinion soon as well.

-Chris

File: README.md [hunk 7]
```diff
@@ -14,17 +14,8 @@ https://www.youtube.com/watch?v=1Tm56JKZcAg
   * `source ~/.virtualenvs/mycroft/bin/activate`
 4. Install the required python libraries
   * `pip install -r requirements.txt`
-5. Add the block below to your mycroft.conf file (`~/.mycroft/mycroft.conf`)
```
Comment (Sujan4k0) [inline README.md:17]: This should be kept as the user has two methods to adjust configurations (CLI or UI), so they both should be listed.

File: README.md [hunk 8]
```diff
@@ -14,17 +14,8 @@ https://www.youtube.com/watch?v=1Tm56JKZcAg
...
-   "SpeedTestSkill": {
-        "host": "",
-        "http_debug": 0,
-        "runs": <runs, recommend 10>
-    }
-
-```
-
-If that file did not already exist (this is the first third party skill you have added), wrap that entire block in { }. The finished file should be valid json. If you have issues, use http://jsonlint.com/ to validate the json.
+5. Change skill options in the mycroft.home if desired (requires reboot)
```
Comment (Sujan4k0) [inline README.md:17]: Need correct URL: https://home.mycroft.ai/#/skill

File: README.md [hunk 9]
```diff
@@ -14,17 +14,8 @@ https://www.youtube.com/watch?v=1Tm56JKZcAg
...
+  * https://www.speedtest.net/speedtest-servers.php
```
Comment (Sujan4k0) [inline README.md:18]: Should provide context to the link. Right now it looks like it's randomly interjected. 

**Example:**
If you wish to dictate a host for the speedtest in the skill's options, here is a list: https://www.speedtest.net/speedtest-servers.php

File: __init__.py [hunk 10]
```diff
@@ -102,24 +106,84 @@ def attempt_three_times(function):
                 raise
             msg = "Caught {error}: {message}. Retrying...".format(
                 error=type(e), message=e.message)
-            logger.warning(msg)
+            LOG.warning(msg)
+
+
+#######################################################################
```
Comment (Sujan4k0) [inline __init__.py:112]: This commenting style is inconsistent with the rest of the file. In addition, there is no indication of where these helpers end. The divider makes it seem that the section is composed of everything below it.

File: __init__.py (hunk 4 above)
Comment (Sujan4k0) [inline __init__.py:115]: After checking out the animation...
- Squiggly lines for the mouth while the speedtest is running is a good idea. 
- The eyes flipping up and down don't add anything to the user experience and make it seem buggy when the eyes are half closed and Mycroft is giving the results at the end. I would remove that and keep them static.

**Review by Sujan4k0 — COMMENTED** [review]
Hey, thanks for contributing! Here are some things that I noticed:

**Pull request closed without merging**